---
title: "Figure 2"
author: "Malte Ruehlemann"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
#bibliography: mylib.bib
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
---

<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r style}
# Set knit global options
library("knitr")
options(digits = 3, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = FALSE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev='png',
               fig.height = 7.5,
               fig.width = 6 * golden_ratio,
               warning = F,
               message =F, 
               comment = '  ',
               dpi = 300,
               cache = FALSE)
library("rmarkdown")

# Pretty outputs
library("BiocStyle")
library("ggthemes") # add ggplot2 themes
library("ggplot2")
theme_set(theme_few(base_size = 18))

# Set seed for reproducibility
set.seed(222)
```

```{r}
#setwd("r_analysis_and_plots")
library(ggplot2)
library(tidyverse)
library(ape)
library(ggtree)
library(compositions)
library(phyloseq)
library(picante)
library(ggrepel)
library(reshape2)
library(cowplot)
library(plotly)
library(TreeDist)
library(phytools)
library(parallel)
library(ggpubr)

interactive=F

load("0_all_preprocessed.Robj")
```



## Results Co-phylogeny analysis

```{r}

genomes_in_cospec = lapply(cospec_results$results, function(x) x$bionj_tree$tip.label) %>% unlist %>% unique

gen_level_cutoff = cospec_results$stats %>% filter(grepl("^g", tax_cons)) 
genomes_in_cospec_gen = lapply((gen_level_cutoff %>% pull(group)), function(x) cospec_results$results[[x]]$bionj_tree$tip.label) %>% unlist %>% unique
sgb_in_cospec = all_genomes %>% filter(genome %in% genomes_in_cospec_gen) %>% pull(GreatApes_95_final) %>% unique

results_cospec = lapply(cospec_results$results, function(x) x$res)  %>% do.call("bind_rows",.) %>% arrange(hommola_p_mean) %>% filter(clade %in% (gen_level_cutoff %>% pull(group)))

results_cospec

genomes_with_cophy = lapply(results_cospec %>% filter(hommola_p_mean<0.05) %>% pull(clade), function(x) cospec_results$results[[x]]$bionj_tree$tip.label) %>% unlist %>% unique()
genomes_with_cophy_gen = genomes_with_cophy[genomes_with_cophy %in% genomes_in_cospec_gen]
sgb_with_cophy = all_genomes %>% filter(genome %in% genomes_with_cophy_gen) %>% pull(GreatApes_95_final) %>% unique


all_genomes %>% filter(!grepl("Manara", genome)) %>% mutate(in_analysis = (genome %in% genomes_in_cospec_gen), cophy = (genome %in% genomes_with_cophy_gen)) %>% filter(in_analysis) %>% mutate(human = grepl("Hs|MG",cluster_97_final)) %>% distinct_at(.var=c("GreatApes_95_final", "human", "cophy")) %>% select(cophy, human) %>% table 

all_cospec_results = results_cospec %>% left_join(gen_level_cutoff, by=c("clade"="group"))
write.table(all_cospec_results, "figures_and_tables/Suppl_Table_S11_Results_Cophylo.tsv", row.names = F, sep="\t")

saveRDS(all_cospec_results, "../../../great_apes_shiny_app/data/all_cospec_results.Rds")
saveRDS(cospec_results, "../../../great_apes_shiny_app/data/cospec_results.Rds")


```


### selected trees for plotting

```{r}

### tree bad

tree_cophy_1 = cospec_results$results$f__UBA932_subtree_75$bionj_tree_rooted %>% drop.tip("outgroup")
tree_cophy_1$node.label = tree_cophy_1$node.label %>% sapply(function(x) strsplit(x, split="/")[[1]][1] %>% as.numeric() )
#if(input$pruned){tree=keep.tip(tree, alldata$results[[c]]$prune_info$genome)}
pp=ggtree(tree_cophy_1, layout=ifelse(is.rooted(tree_cophy_1),"rectangular","equal_angle"), ladderize=T)
pp$data$x = -pp$data$x
pp$data<-left_join(pp$data, cospec_results$genomes, by=c("label"="genome")) %>% left_join(colset %>% mutate(short2=ifelse(grepl("HsG",host), "Hs",host)) %>% select(-host, -host_long, -location), by=c("short2"))
pp_tax_cons = sgb_reps %>% filter(sgb %in% (pp$data %>% filter(isTip) %>% pull(GreatApes_95_final))) %>% pull(genus) %>% unique() %>% paste(collapse=";")
pp_tc = pp$data %>% filter(isTip) %>% select(host_tree, col, shape) %>% distinct()
pp=pp+geom_point(data=pp$data %>% filter(isTip==T),aes(x=x,y=y, shape=host_tree, label=label, col=host_tree, species=species, bin_id=GreatApes_95_final), size=3)+
       scale_color_manual(values=pp_tc$col, breaks=pp_tc$host_tree)+
       scale_shape_manual(values=c(15,16,17,18)[pp_tc$shape],breaks=pp_tc$host_tree) + coord_flip()
pp=pp + geom_point(data=pp$data[!pp$data$isTip & as.numeric(pp$data$label)>90,], size=2) + theme(legend.position = "none") + ggtitle(paste0(pp_tax_cons, " spp."))


### tree no human

tree_cophy_2 = cospec_results$results$f__Paludibacteraceae_subtree_36$bionj_tree_rooted %>% drop.tip("outgroup")
tree_cophy_2$node.label = tree_cophy_2$node.label %>% sapply(function(x) strsplit(x, split="/")[[1]][1] %>% as.numeric() )
#if(input$pruned){tree=keep.tip(tree, alldata$results[[c]]$prune_info$genome)}
pp2=ggtree(tree_cophy_2, layout=ifelse(is.rooted(tree_cophy_2),"rectangular","equal_angle"), ladderize=T)
pp2$data$x = -pp2$data$x
pp2$data<-left_join(pp2$data, cospec_results$genomes, by=c("label"="genome")) %>% left_join(colset %>% mutate(short2=ifelse(grepl("HsG",host), "Hs",host)) %>% select(-host, -host_long, -location), by=c("short2"))
pp2_tax_cons = sgb_reps %>% filter(sgb %in% (pp2$data %>% filter(isTip) %>% pull(GreatApes_95_final))) %>% pull(genus) %>% unique() %>% paste(collapse=";")
pp2_tc = pp2$data %>% filter(isTip) %>% select(host_tree, col, shape) %>% distinct()
pp2=pp2+geom_point(data=pp2$data %>% filter(isTip==T),aes(x=x,y=y, shape=host_tree, label=label, col=host_tree, species=species, bin_id=GreatApes_95_final), size=3)+
       scale_color_manual(values=pp2_tc$col, breaks=pp2_tc$host_tree)+
       scale_shape_manual(values=c(15,16,17,18)[pp2_tc$shape],breaks=pp2_tc$host_tree) + coord_flip()
pp2=pp2 + geom_point(data=pp2$data[!pp2$data$isTip & as.numeric(pp2$data$label)>90,], size=2) + theme(legend.position = "none") + ggtitle(paste0(pp2_tax_cons, " spp."))


### tree human

tree_cophy_3 = cospec_results$results$f__Dialisteraceae_subtree_24$bionj_tree_rooted %>% drop.tip("outgroup")
tree_cophy_3$node.label = tree_cophy_3$node.label %>% sapply(function(x) strsplit(x, split="/")[[1]][1] %>% as.numeric() )
#if(input$pruned){tree=keep.tip(tree, alldata$results[[c]]$prune_info$genome)}
pp3=ggtree(tree_cophy_3, layout=ifelse(is.rooted(tree_cophy_3),"rectangular","equal_angle"), ladderize=T)
pp3$data$x = -pp3$data$x
pp3$data<-left_join(pp3$data, cospec_results$genomes, by=c("label"="genome")) %>% left_join(colset %>% mutate(short2=ifelse(grepl("HsG",host), "Hs",host)) %>% select(-host, -host_long, -location), by=c("short2"))
pp3_tc = pp3$data %>% filter(isTip) %>% select(host_tree, col, shape) %>% distinct()
pp3_tax_cons = sgb_reps %>% filter(sgb %in% (pp3$data %>% filter(isTip) %>% pull(GreatApes_95_final))) %>% pull(genus) %>% unique() %>% paste(collapse=";")
pp3=pp3+geom_point(data=pp3$data %>% filter(isTip==T),aes(x=x,y=y, shape=host_tree, label=label, col=host_tree, species=species, bin_id=GreatApes_95_final), size=3)+
       scale_color_manual(values=pp3_tc$col, breaks=pp3_tc$host_tree)+
       scale_shape_manual(values=c(15,16,17,18)[pp3_tc$shape],breaks=pp3_tc$host_tree) + coord_flip()
pp3=pp3 + geom_point(data=pp3$data[!pp3$data$isTip & as.numeric(pp3$data$label)>90,], size=2) + theme(legend.position = "none") + ggtitle(paste0(pp3_tax_cons, " spp."))


### tree archaea

tree_cophy_4 = cospec_results$results$f__Methanomethylophilaceae_subtree_24$bionj_tree_rooted %>% drop.tip("outgroup")
tree_cophy_4$node.label = tree_cophy_4$node.label %>% sapply(function(x) strsplit(x, split="/")[[1]][1] %>% as.numeric() )
#if(input$pruned){tree=keep.tip(tree, alldata$results[[c]]$prune_info$genome)}
pp4=ggtree(tree_cophy_4, layout=ifelse(is.rooted(tree_cophy_4),"rectangular","equal_angle"), ladderize=T)
pp4$data$x = -pp4$data$x
pp4$data<-left_join(pp4$data, cospec_results$genomes, by=c("label"="genome")) %>% left_join(colset %>% mutate(short2=ifelse(grepl("HsG",host), "Hs",host)) %>% select(-host, -host_long, -location), by=c("short2"))
pp4_tc = pp4$data %>% filter(isTip) %>% select(host_tree, col, shape) %>% distinct()
pp4_tax_cons = sgb_reps %>% filter(sgb %in% (pp4$data %>% filter(isTip) %>% pull(GreatApes_95_final))) %>% pull(genus) %>% unique() %>% paste(collapse=";")
pp4=pp4+geom_point(data=pp4$data %>% filter(isTip==T),aes(x=x,y=y, shape=host_tree, label=label, col=host_tree, species=species, bin_id=GreatApes_95_final), size=3)+
       scale_color_manual(values=pp4_tc$col, breaks=pp4_tc$host_tree)+
       scale_shape_manual(values=c(15,16,17,18)[pp4_tc$shape],breaks=pp4_tc$host_tree) + coord_flip()
pp4=pp4 + geom_point(data=pp4$data[!pp4$data$isTip & as.numeric(pp4$data$label)>90,], size=2) + theme(legend.position = "none") + ggtitle(paste0(pp4_tax_cons, " spp."))


cophy_tree_set = plot_grid(NULL, pp,pp2,pp3,pp4,get_legend(pp + theme(legend.position = "top") + guides(color=guides(colour = guide_legend(ncol = 2, title=""), shape= guide_legend(ncol = 2, title="")))), rel_heights = c(0.05,0.2,0.2,0.2,0.2,0.2), ncol=1)


```

### cophylogeny enrichment

```{r}
fam_cophy_stats = all_genomes %>% filter(GreatApes_95_final %in% sgb_in_cospec) %>% mutate(cophy=(genome %in% genomes_with_cophy_gen)) %>% select(GreatApes_95_final, cophy) %>% left_join(tax2) %>% distinct %>% select(family, cophy) %>% table %>% data.frame(stringsAsFactors = F) 

fisher_out = lapply(as.character(unique(fam_cophy_stats$family)), function(ff){
  fdf=fam_cophy_stats %>% mutate(this_fam = family==ff) %>% group_by(this_fam, cophy) %>% summarize(n=sum(Freq)) %>% reshape2::dcast(this_fam ~ cophy, value.var = "n") %>% column_to_rownames("this_fam")
 p=(fdf %>% fisher.test())$p.value
 return(data.frame(fam=ff, p_fisher=p, cosp_rat=fdf[2,2]/sum(fdf[2,])))
}) %>% do.call("bind_rows",.)

cophy_stats2 = fisher_out %>% left_join(fam_cophy_stats %>% group_by(family) %>% summarize(tot=sum(Freq)), by=c("fam"="family")) %>% arrange(p_fisher) %>% mutate(qval=p.adjust(p_fisher, "fdr")) %>% left_join(sgb_reps %>% select(phylum2,fam=family) %>% distinct()) %>% mutate(fam = fct_reorder(fam, -cosp_rat))

cophy_mean = cophy_stats2 %>% filter(tot>=10) %>% pull(cosp_rat) %>% mean

cophy_enrich_plot = cophy_stats2 %>% filter(tot>=10, qval<0.05) %>% ggplot(aes(y=cosp_rat*100, x=fam, fill=phylum2)) + geom_col() + geom_hline(yintercept = cophy_mean*100, lty=2) + scale_fill_manual(values=phylumcols_df$col, breaks=phylumcols_df$phylum) + ylab("Cophylo-Ratio (%)") + xlab("")+ theme(axis.text.x = element_text(angle=45, hjust=1), legend.position = "none") + scale_y_continuous(expand=c(.01,0))

cophy_stats2
write.table(cophy_stats2, "figures_and_tables/Suppl_Table_S12_Cophylo_Families.tsv", row.names = F, sep="\t")


saveRDS(cophy_stats2, "../../../great_apes_shiny_app/data/family_cophy_ratios.Rds")
```

### per-sample sgb co-phylo ratio


```{r}
coph_plot = sgb_long %>% filter(variable %in% (all_genomes %>% filter(genome %in% genomes_with_cophy_gen) %>% pull(GreatApes_95_final))) %>% group_by(sample, host_long) %>% summarize(totabu=sum(value)/10000) %>% data.frame %>% ggplot(aes(x=host_long,y=totabu, fill=host_long)) + geom_boxplot(outlier.shape = NA) + geom_jitter(height=0, width=.2) + scale_fill_manual(values=colset$col, breaks = colset$host_long) + theme(legend.position = "none", axis.text.x = element_text(angle=45, hjust=1)) + ylab("Abundance in Co-Phylo SGBs") + xlab("") 

coph_df=sgb_long %>% filter(variable %in% (all_genomes %>% filter(genome %in% genomes_with_cophy_gen) %>% pull(GreatApes_95_final))) %>% group_by(sample, host_long) %>% summarize(totabu=sum(value)/10000) %>% data.frame 

coph_df_a=sgb_long %>% filter(variable %in% (all_genomes %>% filter(genome %in% genomes_with_cophy_gen) %>% pull(GreatApes_95_final))) %>% group_by(sample, host_long) %>% summarize(totabu=sum(value)/10000) %>% data.frame 

coph_df_a %>% ggplot(aes(x=host_long,y=totabu, fill=host_long)) + geom_boxplot(outlier.shape = NA) + geom_jitter(height=0, width=.2) + scale_fill_manual(values=colset$col, breaks = colset$host_long) + theme(legend.position = "none", axis.text.x = element_text(angle=45, hjust=1)) + ylab("Abundance in Co-Phylo SGBs") + xlab("") 



coph_plot =  sgb_long %>% filter(variable %in% sgb_in_cospec) %>% mutate(pres = value > 0, cosp = variable %in% sgb_with_cophy) %>% filter(pres) %>% group_by(sample, host_long) %>% summarise(cophy_rat = mean(cosp)) %>%  ggplot(aes(x=host_long,y=cophy_rat*100, fill=host_long)) + geom_boxplot(outlier.shape = NA) + geom_jitter(height=0, width=.2) + scale_fill_manual(values=colset$col, breaks = colset$host_long) + theme(legend.position = "none", axis.text.x = element_text(angle=45, hjust=1)) + ylab("Prop. of Co-Phylo SGBs (%)") + xlab("") + scale_y_continuous(expand=c(.01,0))

sgb_long %>% filter(variable %in% sgb_in_cospec) %>% mutate(pres = value > 0)  %>% filter(pres) %>% group_by(sample, host_long) %>% summarise(in_cophy = n())  %>%  ggplot(aes(x=host_long,y=in_cophy, fill=host_long)) + geom_boxplot(outlier.shape = NA) + geom_jitter(height=0, width=.2) + scale_fill_manual(values=colset$col, breaks = colset$host_long) + theme(legend.position = "none", axis.text.x = element_text(angle=45, hjust=1)) + ylab("Prop. of Co-Phylo SGBs (%)") + xlab("") + scale_y_continuous(expand=c(.01,0))


ggpubr::compare_means(data=coph_plot$data, cophy_rat ~ host_long)


saveRDS(coph_plot$data, "../../../great_apes_shiny_app/data/sample_coph_plot.Rds")


```

### sporulation genes

```{r}
spore_kos = KO_NAMES %>% filter(grepl("sporulation|spore",annot)) %>% pull(ko)

this_to_sgb %>% filter(ogs %in% gsub("ko:","",spore_kos), GreatApes_95_final %in% sgb_in_cospec) %>% select(-host_group) %>% distinct() %>% group_by(GreatApes_95_final) %>% summarise(spore_count = n()) %>% left_join(tax2)  %>% group_by(phylum, family, genus) %>% summarise(mean_spore_count = mean(spore_count)) %>% ungroup() %>% left_join(cophy_stats2, by=c("family"="fam")) %>% filter(tot>=10) %>% mutate(fam=as.factor(family)) %>% lm(cosp_rat ~ mean_spore_count, data=.) %>% summary()

spore_plot = this_to_sgb %>% filter(ogs %in% gsub("ko:","",spore_kos), GreatApes_95_final %in% sgb_in_cospec) %>% select(-host_group) %>% distinct() %>% group_by(GreatApes_95_final) %>% summarise(spore_count = n()) %>% left_join(tax2)  %>% group_by(phylum, family, genus) %>% summarise(mean_spore_count = mean(spore_count)) %>% ungroup() %>% left_join(cophy_stats2, by=c("family"="fam")) %>% filter(tot>=10) %>% mutate(fam=as.factor(family)) %>% ggplot(aes(x=mean_spore_count, y=cosp_rat*100)) + geom_hline(yintercept = cophy_mean*100, lty=2) + geom_point(aes(col=phylum2)) + stat_smooth(method = "lm", col="red", se = F) + scale_color_manual(values=phylumcols_df$col, breaks=phylumcols_df$phylum) + scale_y_continuous(expand=c(.01,0), name = "Cophylo-Ratio (%)") + scale_x_continuous(name="# of Sporulation Genes")

sgb_spore_kos_prev = this_to_sgb %>% filter(ogs %in% gsub("ko:","",spore_kos), GreatApes_95_final %in% sgb_in_cospec) %>% select(-host_group) %>% distinct() %>% mutate(count=1) %>% dcast(GreatApes_95_final ~ ogs, value.var="count", fill=0)

write.table(sgb_spore_kos_prev, "figures_and_tables/Suppl_Table_S13_SGB_Sporulation_Genes.tsv", row.names = F, sep="\t")

saveRDS(spore_plot$data, "../../../great_apes_shiny_app/data/sgb_spore_plot.Rds")


```

### Bifido tree

```{r}
library(dendextend)
 library(phangorn)


info_sub = gyrb_info %>% filter(genus=="g__Bifidobacterium", prot_id %in% bifido_gyrb_tree$tip.label) %>% filter(!is.na(host_tree)) %>% group_by(species, host_tree) %>% sample_n(1) %>% ungroup %>% mutate(dup = duplicated(GreatApes_95_final)) %>% mutate(label = paste0(species, " (",host_group,")")) %>% mutate(pre_lab = ifelse(dup, paste0(GreatApes_95_final,"_",host_group), GreatApes_95_final))

bifido_gyrb_tree_sub = bifido_gyrb_tree %>% keep.tip(c("outgroup",  info_sub %>% pull(prot_id)))

bifido_gyrb_tree_sub$tip.label = data.frame(info_sub)[match(bifido_gyrb_tree_sub$tip.label, info_sub$prot_id), "label"] %>%  ifelse(bifido_gyrb_tree_sub$tip.label == "outgroup", "outgroup",.) 

bifido_marker_tree2 = add.tips(tree = bifido_marker_tree, where = info_sub %>% filter(dup) %>% pull(GreatApes_95_final), tips = info_sub %>% filter(dup) %>% pull(pre_lab) )

bifido_marker_tree2$tip.label = data.frame(info_sub)[match(bifido_marker_tree2$tip.label, info_sub$pre_lab), "label"] %>% ifelse(bifido_marker_tree2$tip.label == "outgroup", "outgroup",.) 
#bifido_marker_tree = bifido_marker_tree %>% keep.tip(., bifido_gyrb_tree_sub$tip.label)

dend_marker = chronos(bifido_marker_tree2 %>% keep.tip(bifido_gyrb_tree_sub$tip.label) %>% ape::root(phy=., outgroup="outgroup",resolve.root=T) )
dend_gyrb = chronos(bifido_gyrb_tree_sub) 

tanglegram(dend_marker, dend_gyrb, margin_inner=20)


tg=tanglegram(dend_marker, dend_gyrb, sort = TRUE, common_subtrees_color_lines = T, highlight_distinct_edges  = FALSE, highlight_branches_lwd = FALSE, margin_inner=0, axes=F,columns_width=c(1,1,1), lab.cex = NA, lwd=2) 
tg2 = tg %>% untangle(method = "step2side") 

png("gyrb_tangle.png", width = 14, height=20, units = "cm", res = 300)
tg2 %>% plot(common_subtrees_color_lines = T, highlight_distinct_edges  = FALSE, highlight_branches_lwd = FALSE, margin_inner=0.1, axes=F,columns_width=c(1,1,1), lab.cex = 2, lwd=2, edge.lwd=2, main_left = "GTDB Markers", main_right="gyrB")
dev.off()

bifido_colors = bind_rows(bind_rows(info_sub[match(bifido_gyrb_tree_sub$tip.label, info_sub$label),] %>% mutate(host = ifelse(host=="HsUHGG", "HsG", host)) %>% left_join(colset)) %>% select(label, GreatApes_95_final, prot_id, host, host_tree, col, species, shape) %>% filter(!is.na(host)), data.frame(stringsAsFactors = F,label="outgroup", GreatApes_95_final="outgroup",prot_id="outgroup",host="outgroup",host_tree="outgroup",col="grey", species="outgroup", shape=1) %>% group_by(host_tree))


bifido_marker = tg2$dend1 %>% as.phylo() %>% ape::root(phy=., outgroup="outgroup",resolve.root=T) %>% ggtree(, size=.8)  + theme_nothing() + scale_y_continuous(limits=c(0.5, 13)) + geom_tippoint(size=3, aes(color=label, shape=label)) + scale_color_manual(values=bifido_colors$col, breaks = bifido_colors$label) 
#bifido_marker$data = bifido_marker$data %>% mutate(x=ifelse(label=="outgroup", bifido_marker$data %>% filter(isTip, label!="outgroup") %>% pull(x) %>% max(),x))
bifido_marker = bifido_marker + geom_segment(data = . %>% filter(isTip), aes(x=x,xend= bifido_marker$data %>% filter(isTip, label!="outgroup") %>% pull(x) %>% max(), y=y, yend=y), lty=2, size=.2) + scale_shape_manual(values=c(15,16,17,18)[bifido_colors$shape], breaks=bifido_colors$label) + geom_point(data = . %>% mutate(label2 = gsub("/[0-9]+","",label)) %>% filter(!isTip, as.numeric(label2) > 80))


bifido_gyrb = tg2$dend2 %>% as.phylo() %>% root(outgroup="outgroup", resolve.root=T) %>% ggtree(size=.8)  + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 13)) + geom_tippoint(aes(color=label, shape=label), size=3)  
#bifido_gyrb$data = bifido_gyrb$data %>% mutate(x=ifelse(label=="outgroup", bifido_gyrb$data %>% filter(isTip, label!="outgroup") %>% pull(x) %>% max(),x)) 
bifido_gyrb = bifido_gyrb + scale_color_manual(values=bifido_colors$col, breaks = bifido_colors$label) + geom_segment(data = . %>% filter(isTip), aes(x=x,xend= bifido_gyrb$data %>% filter(isTip, label!="outgroup") %>% pull(x) %>% max(), y=y, yend=y), lty=2, size=.2) + scale_shape_manual(values=c(15,16,17,18)[bifido_colors$shape], breaks=bifido_colors$label) + geom_point(data = . %>% filter(!isTip, as.numeric(label) > 80))

conn2 = bifido_marker$data %>% filter(isTip) %>% select(label,y_markers=y) %>% left_join(bifido_gyrb$data %>% filter(isTip) %>% select(label,y_gyrb=y)) %>% ggplot(aes(x=1, xend=2, y=y_markers, yend=y_gyrb)) + geom_segment(size=.8, lty=2, aes(color=label)) + theme_nothing() + scale_y_continuous(limits=c(0.5, 13)) + scale_color_manual(values=bifido_colors$col, breaks = bifido_colors$label)

labelplot = bifido_marker$data %>% filter(isTip) %>% ggplot(aes(x=0, y=y, col=label)) + theme_nothing() + scale_shape_manual(values=c(16,17,18)[bifido_colors$shape], breaks = bifido_colors$label) + scale_color_manual(values=bifido_colors$col, breaks = bifido_colors$label) + geom_text(aes(label=label), hjust=0) + scale_x_continuous(limits=c(0,1.5)) + scale_y_continuous(limits=c(0.5, 13))

bifido_plot = plot_grid(NULL,
  bifido_marker + geom_text(data=data.frame(x=0,y=13,label="GTDB Markers"), hjust=0, aes(label=label)) , 
 labelplot,  conn2,
  bifido_gyrb + geom_text(data=data.frame(x=0,y=13,label="gyrB"), hjust=1, aes(label=label)), NULL,
   ncol=6,
  rel_widths = c(.025,0.15,0.3,0.1,0.15,.025)
)

# bifido_plot = plot_grid(
#   bifido_marker + geom_text(data=data.frame(x=0,y=8,label="GTDB Markers"), hjust=0, aes(label=label)) , 
#   conn2, 
#   bifido_gyrb + geom_text(data=data.frame(x=0,y=8,label="gyrB"), hjust=1, aes(label=label)) ,
#   get_legend(bifido_marker + theme(legend.position = "left", legend.justification = "center") +  guides(colour = guide_legend(col = 1, title=""), shape = guide_legend(col = 1, title=""))), ncol=4,
#   rel_widths = c(0.2,0.2,0.2,0.4)
# )

bifido_moeller_info = data.frame(tip=bifido_moeller$tip.label, sample = bifido_moeller$tip.label %>% gsub("Spades_","",.) %>% lapply(., function(x) strsplit(x, split="_")[[1]][1]) %>% unlist %>% gsub("[0-9]{2,}$","",.), stringsAsFactors = F) %>% left_join(sample_meta) %>% mutate(hostnew=ifelse(is.na(host), sample, host)) %>% mutate(hostnew=ifelse(grepl("^Pt",hostnew), "Chimp",hostnew), hostnew=ifelse(grepl("^Hs|^MGYG",hostnew), "Human",hostnew), hostnew=ifelse(grepl("^Pp",hostnew), "Bonobo",hostnew), hostnew=ifelse(grepl("^G|^SRR",hostnew), "Gorilla",hostnew)) %>% mutate(group=ifelse(sample %in% c("Human","Gorilla","Bonobo","Chimp"),"Moeller","New")) %>% select(tip, hostnew,group)

gg_moe = bifido_moeller %>% ggtree()
gg_moe$data = gg_moe$data %>% left_join(bifido_moeller_info, by=c("label"="tip"))
gg_moe + geom_tippoint(aes(col=hostnew, shape=group)) + scale_x_reverse() + scale_color_manual(values=c("red","orange","darkgreen","lightblue","grey")) + scale_shape_manual(values=c(21,16))

```

```{r, fig.height=4, fig.width=15}
top_right= align_plots(cophy_enrich_plot,
        coph_plot, align = "h")

plot3=plot_grid(
  cophy_tree_set, NULL,
    plot_grid(
        plot_grid(top_right[[1]],NULL, top_right[[2]], labels=c("b","c"), ncol=3, rel_widths = c(.45,.025,.45)
        ),
        plot_grid(NULL, spore_plot + theme(legend.position = "none"), get_legend(spore_plot  + theme(legend.position = "left") + guides(color=guide_legend(title="Phlyum", ncol=2, label.vjust=0))), rel_widths = c(.025,.4,.6), labels=c("d",""), ncol=3),
      bifido_plot, ncol=1, rel_heights = c(.5,.25,.25), labels=c("","", "e")),
  rel_widths = c(.3,.1,.7), labels=c("a","",""), ncol=3
)


plot3

ggsave(filename = "figures_and_tables/Figure_3.png", plot3, height=12, width=15, bg="white")

```
