---
title: "Figure 1; Figure S2"
author: "Malte Ruehlemann"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
#bibliography: mylib.bib
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
---

<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r style}
# Set knit global options
library("knitr")
options(digits = 3, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = FALSE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev='png',
               fig.height = 7.5,
               fig.width = 6 * golden_ratio,
               warning = F,
               message =F, 
               comment = '  ',
               dpi = 300,
               cache = FALSE)
library("rmarkdown")

# Pretty outputs
library("BiocStyle")
library("ggthemes") # add ggplot2 themes
library("ggplot2")
theme_set(theme_few(base_size = 18))

# Set seed for reproducibility
set.seed(222)
```

```{r}
#setwd("r_analysis_and_plots")
library(ggplot2)
library(tidyverse)
library(ape)
library(ggtree)
library(compositions)
library(phyloseq)
library(picante)
library(ggrepel)
library(reshape2)
library(cowplot)
library(plotly)
library(TreeDist)
library(phytools)
library(parallel)
library(ggpubr)

interactive=F
intermediatedata_folder="~/Desktop/Projects/190712_GreatApes/220303_Analysis_New/intermediate_data/"
setwd("~/Desktop/Projects/190712_GreatApes/220616_dataprocessing/greatapes_mgx_scripts/r_analysis_and_plots//")
load("0_all_preprocessed.Robj")
```

```{r}

sra1=read.table("https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJNA692042&result=read_run&fields=study_accession,sample_accession,experiment_accession,run_accession,tax_id,scientific_name,library_strategy,fastq_bytes,fastq_ftp,submitted_ftp,sra_ftp,sample_alias&format=tsv&download=true&limit=0", sep="\t", head=T, stringsAsFactors = F) %>% select(sample=sample_alias, run_accession, sample_accession, study_accession) %>% mutate(sample=paste0(sample,"-L1"))

sra2=read.table("https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJNA539933&result=read_run&fields=study_accession,sample_accession,experiment_accession,run_accession,tax_id,scientific_name,library_strategy,fastq_bytes,fastq_ftp,submitted_ftp,sra_ftp,sample_alias&format=tsv&download=true&limit=0", sep="\t", head=T, stringsAsFactors = F) %>% mutate(sample=run_accession) %>% select(sample, run_accession, sample_accession, study_accession)

sra3=read.table("https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJNA491335&result=read_run&fields=study_accession,sample_accession,experiment_accession,run_accession,tax_id,scientific_name,library_strategy,fastq_bytes,fastq_ftp,submitted_ftp,sra_ftp,sample_alias&format=tsv&download=true&limit=0", sep="\t", head=T, stringsAsFactors = F) %>% select(sample=sample_alias, run_accession, sample_accession, study_accession) %>% mutate(sample=paste0("SID",sample))

write.table(sample_meta %>% select(-shape, -gid) %>% left_join(bind_rows(sra1,sra2,sra3)) %>% filter(!is.na(run_accession)), "figures_and_tables/Suppl_Table_S1_Sample_Metadata.tsv", sep="\t", row.names = F, quote=F)


write.table(all_genomes_final %>% filter(!(grepl("MGYG",genome) & cluster_95_final!=genome)), "figures_and_tables/Suppl_Table_S2_All_MAGs.tsv", sep="\t", row.names = F, quote=F)

```



### Fig1 - Panel A

```{r, fig.height=11.5, fig.width=6.5}
#####

### calculate phylogenetic diversity of rarefied SGB abundances
salmon_rare_pd_file = paste0(intermediatedata_folder, "salmon_rare_pd.Rds")

if(file.exists(salmon_rare_pd_file)){
  salmon_rare_pd=readRDS(salmon_rare_pd_file)
}else{
  salmon_rare_pd = picante::pd(salmon_rare %>% mutate(sid = paste0(sample,"_",depth,"_",i)) %>% dcast(sid ~ bin, value.var="tpm", fill=0) %>% column_to_rownames("sid"), tree = tree2, include.root = T)
  saveRDS(salmon_rare_pd, salmon_rare_pd_file)
}

### used PD at 1 Million Reads
phlyodiv=salmon_rare_pd %>% rownames_to_column("sid") %>% separate(sid, into = c("sample","depth","i"), sep="_") %>% mutate(depth=as.numeric(depth)) %>% filter(depth==1000000)

fig1_panel_a = phlyodiv %>% group_by(sample) %>% summarise(PD=mean(PD)) %>% left_join(sample_meta) %>% mutate(host2=ifelse(host_tree=="Homo_sapiens", host, gsub("[0-9]","",host))) %>% ggplot(aes(x=host2, y=PD, fill=host2, shape=host, group=host2)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width=.2, size=1) + theme(axis.text.x = element_text(angle=45, hjust=1), axis.title = element_text(size=10)) + scale_fill_manual(values=colset$col, breaks=colset$host) + scale_y_continuous(limits=c(0,75), expand=c(0,0)) + theme(legend.position = "none", axis.text.x = element_blank(), axis.ticks.x=element_blank()) + ylab("PhyloDiv @ 1 Mio Reads") + xlab("Host") + scale_shape_manual(values=c(22,21,24,23,21)[colset$shape], breaks=colset$host) #+ ggpubr::stat_compare_means(comparisons = combn(colset$host_long, 2, simplify = F), hide.ns = T)

pd_sign = fig1_panel_a$data %>%  ggpubr::compare_means(data=., PD ~ host2)

pd_nhp_human = fig1_panel_a$data %>% mutate(h_vs_nhp = (host2 %in% c("HsA1","HsA2","HsG","HsD"))) %>% ggpubr::compare_means(data=., PD ~ h_vs_nhp)


```


# Beta diversity plots - Panel B and Fig S2

```{r, fig.height=3}

write.table(sgb_abundance %>% rownames_to_column("sample"), "figures_and_tables/Suppl_Table_S3_SGB_Abundances.tsv", sep="\t", row.names = F, quote=F)

colset=colset %>% arrange(host_long)
## clr transformation of genus level abundances
gen = gen[!rownames(gen) %in% sample_exclude,]
sgb_abundance = sgb_abundance[!rownames(sgb_abundance) %in% sample_exclude,]
sgb_long = sgb_long %>% filter(!sample %in% sample_exclude)

gen_clr=compositions::clr(gen+1)
gen_clr[gen_clr<0] = 0

## unifrac weighted 
uf=UniFrac(phyloseq(otu_table(sgb_abundance, taxa_are_rows = F), tree=tree2),weighted = T, normalized = T)
ufplot1=cmdscale(uf) %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("UniFrac weighted PCoA1")  + ylab("UniFrac weighted PCoA2") 

## unifrac unweighted
uf_uw=UniFrac(phyloseq(otu_table(sgb_abundance, taxa_are_rows = F), tree=tree2),weighted = F, normalized = T)
ufplot2=cmdscale(uf_uw) %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("UniFrac unweighted PCoA1")  + ylab("UniFrac unweighted PCoA2") 

fig1_panel_b = ufplot2


## clr
clr_dist = gen_clr %>% dist
clrplot=cmdscale(clr_dist) %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("CLR PCoA1")  + ylab("CLR PCoA2") 

## jaccard
jac_dist = gen %>% vegdist(binary = T)
jacplot=cmdscale(jac_dist) %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("Jaccard PCoA1")  + ylab("Jaccard PCoA2") 

## gene abundance
func_abu_long = sgb_long %>% filter(value>0) %>% mutate(host_group=substr(host,1,1)) %>% select(sample, GreatApes_95_final=variable, host_group, value) %>% left_join(pangenome_all) %>% group_by(sample, ogs) %>% summarize(abundance=sum(value)) %>% filter(!is.na(ogs))
func_abu = reshape2::dcast(sample ~ ogs, value.var="abundance", fill=0, data=func_abu_long) %>% data.frame %>% column_to_rownames("sample")
# func_clr=compositions::clr(func_abu+1)
# func_clr[func_clr<0] = 0
func_abu_log = log(func_abu+1)
funcplot1 = func_abu_log %>% dist %>% cmdscale() %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("Function Abundance PCoA1")  + ylab("Function Abundance PCoA2") 
#fig1_panel_c = funcplot1


func_pa = reshape2::dcast(sample ~ ogs, value.var="abundance", fill=0, data=func_abu_long %>% mutate(abundance=ifelse(abundance>0,1,0))) %>% data.frame %>% column_to_rownames("sample")
funcplot2 = func_pa %>% vegdist(binary = T) %>% cmdscale() %>% data.frame() %>% rownames_to_column("sample") %>% left_join(sample_meta) %>% ggplot(aes(x=X1, y=X2, col=host_long, shape=host_long)) + geom_point(size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host_long) + theme_bw() + theme(panel.grid = element_blank()) + xlab("Function P/A PCoA1")  + ylab("Function P/A PCoA2") 


```

# Phylosymbiosis analysis

## Random trees for phylosymbiosis p-value calculation

```{r}
### Random tree generation for RF calculations

random_trees_100k_rf_file = paste0(intermediatedata_folder, "random_trees_100k_rf.Rds")

if(file.exists(random_trees_100k_rf_file)){
  random_trees_100k_rf=readRDS(random_trees_100k_rf_file)
}else{
random_trees_100k_rf=list()
random_trees_100k_rf[["ufuw"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
random_trees_100k_rf[["ufw"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
random_trees_100k_rf[["clr"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
random_trees_100k_rf[["jac"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
random_trees_100k_rf[["func"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
random_trees_100k_rf[["funcpa"]] = lapply(1:100000, function(x) phangorn::RF.dist(host_tree, rtree(7, rooted=F, tip.label = host_tree$tip.label)))
saveRDS(random_trees_100k_rf, random_trees_100k_rf_file)
}

```

## Full dataset

```{r, fig.height=3}
###### calculate cumulative per-host abundances and generate jacknifed trees

sgb_abu_host = sgb_long %>% filter(!is.na(host)) %>% mutate(host2 = ifelse(grepl("Hs",host),"Hs",gsub("[0-9]+","",host))) %>% group_by(host2, variable) %>% summarize(value=sum(as.integer(value))) %>% reshape2::dcast(host2 ~ variable, value.var="value", fill=0) %>% data.frame %>% column_to_rownames("host2")

gen_abu_host = gen %>% data.frame %>% rownames_to_column("sample") %>% reshape2::melt(.) %>% left_join(sample_meta) %>% mutate(value=ifelse(value<0,0,value)) %>% filter(!is.na(host)) %>% mutate(host2 = ifelse(grepl("Hs",host),"Hs",gsub("[0-9]+","",host))) %>% group_by(host2, variable) %>% summarize(value=sum(as.integer(value))) %>% reshape2::dcast(host2 ~ variable, value.var="value", fill=0) %>% data.frame %>% column_to_rownames("host2")

func_abu_host = func_abu %>% data.frame %>% rownames_to_column("sample") %>% reshape2::melt(.) %>% left_join(sample_meta) %>% mutate(value=ifelse(value<0,0,value)) %>% filter(!is.na(host)) %>% mutate(host2 = ifelse(grepl("Hs",host),"Hs",gsub("[0-9]+","",host))) %>% group_by(host2, variable) %>% summarize(value=sum(as.integer(value))) %>% reshape2::dcast(host2 ~ variable, value.var="value", fill=0) %>% data.frame %>% column_to_rownames("host2")
func_abu_host_log = log(func_abu_host+1)

func_abu_host_pa = func_abu_host
func_abu_host_pa[func_abu_host_pa>0] = 1

gen_abu_host_clr = clr(gen_abu_host+1)
gen_abu_host_clr[gen_abu_host_clr<0]=0

### Jacknife trees

jackknife_trees_file = paste0(intermediatedata_folder, "jackknife_trees.Rds")

if(file.exists(jackknife_trees_file)){
  jackknife_trees=readRDS(jackknife_trees_file)
}else{
jackknife_trees = mclapply(1:1000, function(x){
  set.seed(x)
  trr = vegan::rrarefy(sgb_abu_host, sample=1000000)
  trr_gen = (vegan::rrarefy(gen_abu_host, sample=1000000))
  trr_gen_clr = clr(trr_gen+1)
  trr_gen_clr[trr_gen_clr<0] = 0
  trr_func = trr %>% data.frame() %>% rownames_to_column("host") %>% reshape2::melt(.) %>% mutate(host_group = substr(host,1,1)) %>% filter(value>0) %>% left_join(pangenome_all, by=c("variable"="GreatApes_95_final")) %>% group_by(host, ogs) %>% summarize(abundance=sum(value)) %>% filter(!is.na(ogs)) %>% dcast(host ~ ogs, value.var="abundance", fill=0) %>% column_to_rownames("host")
  trr_func_log=  log(trr_func+1)
  trr_func_pa = trr_func
  trr_func_pa[trr_func_pa>0] = 1
  ps_trr = phyloseq(otu_table=otu_table(trr, taxa_are_rows = F), tree=tree2)
  return(list(clr_tree = trr_gen_clr %>% dist %>% hclust(method="average") %>% as.phylo,
              jac_tree = trr_gen %>% vegdist(binary = T) %>% hclust(method="average") %>% as.phylo,
              ufuw_tree =ps_trr %>% UniFrac(weighted = F, normalized = F) %>% hclust(method="average") %>% as.phylo,
              ufw_tree = ps_trr %>% UniFrac(weighted = T, normalized = F) %>% hclust(method="average") %>% as.phylo,
              func_tree = trr_func_log %>% dist %>% hclust(method="average") %>% as.phylo,
              funcpa_tree = trr_func_pa %>% vegdist(binary = T) %>% hclust(method="average") %>% as.phylo))
}, mc.cores = 10)
saveRDS(jackknife_trees, jackknife_trees_file)
}

### Jackknife consensus trees
subset = sample_meta %>% filter(sample %in% rownames(sgb_abundance)) %>% group_by(host_tree) %>% slice_sample(n=1) %>% mutate(host3 =ifelse(grepl("Hs",host),"Hs", gsub("[0-9]+","",host))) %>% select(host_tree, host3)

tree_clr_host_full = gen_abu_host_clr  %>% dist %>% hclust(method="average") %>% as.phylo %>% unroot
tree_clr_host_full$node.label = prop.clades(tree_clr_host_full, lapply(jackknife_trees, function(x) x$clr_tree %>% unroot )  %>% data.table::setattr("class", "multiPhylo"))
tree_clr_host_full$tip.label = subset[match(tree_clr_host_full$tip.label, subset$host3),1] %>% unlist %>% as.vector

tree_ufuw_host_full = (sgb_abu_host/rowSums(sgb_abu_host)) %>% otu_table(., taxa_are_rows = F) %>% phyloseq(otu_table=., tree=tree2) %>% UniFrac(weighted = F, normalized = T) %>% hclust(method="average") %>% as.phylo %>% unroot
tree_ufuw_host_full$node.label = prop.clades(tree_ufuw_host_full, lapply(jackknife_trees, function(x) x$ufuw_tree %>% unroot )  %>% data.table::setattr("class", "multiPhylo"))
tree_ufuw_host_full$tip.label = subset[match(tree_ufuw_host_full$tip.label, subset$host3),1] %>% unlist%>% as.vector

tree_ufw_host_full = (sgb_abu_host/rowSums(sgb_abu_host)) %>% otu_table(., taxa_are_rows = F) %>% phyloseq(otu_table=., tree=tree2) %>% UniFrac(weighted = T, normalized = T) %>% hclust(method="average") %>% as.phylo %>% unroot
tree_ufw_host_full$node.label = prop.clades(tree_ufw_host_full, lapply(jackknife_trees, function(x) x$ufw_tree %>% unroot )  %>% data.table::setattr("class", "multiPhylo"))
tree_ufw_host_full$tip.label = subset[match(tree_ufw_host_full$tip.label, subset$host3),1] %>% unlist%>% as.vector

tree_jac_host_full = gen_abu_host %>% vegdist(binary=T) %>% hclust(method="average") %>% as.phylo %>% unroot
tree_jac_host_full$node.label = prop.clades(tree_jac_host_full, lapply(jackknife_trees, function(x) x$jac_tree  %>% unroot) %>% unclass %>% data.table::setattr("class", "multiPhylo"))
tree_jac_host_full$tip.label = subset[match(tree_jac_host_full$tip.label, subset$host3),1] %>% unlist%>% as.vector

tree_func_host_full = func_abu_host_log %>% dist %>% hclust(method="average") %>% as.phylo %>% unroot
tree_func_host_full$node.label = prop.clades(tree_func_host_full, lapply(jackknife_trees, function(x) x$func_tree  %>% unroot) %>% unclass %>% data.table::setattr("class", "multiPhylo"))
tree_func_host_full$tip.label = subset[match(tree_func_host_full$tip.label, subset$host3),1] %>% unlist%>% as.vector

tree_funcpa_host_full = func_abu_host_pa %>% vegdist(binary=T) %>% hclust(method="average") %>% as.phylo %>% unroot
tree_funcpa_host_full$node.label = prop.clades(tree_funcpa_host_full, lapply(jackknife_trees, function(x) x$func_tree  %>% unroot) %>% unclass %>% data.table::setattr("class", "multiPhylo"))
tree_funcpa_host_full$tip.label = subset[match(tree_funcpa_host_full$tip.label, subset$host3),1] %>% unlist%>% as.vector


p_phylosym=data.frame(
clr_full=(unlist(random_trees_100k_rf["clr"]) <= phangorn::RF.dist(host_tree,tree_clr_host_full)) %>% mean,
ufw_full=(unlist(random_trees_100k_rf["ufw"]) <= phangorn::RF.dist(host_tree,tree_ufw_host_full)) %>% mean,
ufuw_full=(unlist(random_trees_100k_rf["ufuw"]) <= phangorn::RF.dist(host_tree,tree_ufuw_host_full)) %>% mean,
jac_full=(unlist(random_trees_100k_rf["jac"]) <= phangorn::RF.dist(host_tree,tree_jac_host_full)) %>% mean,
func_full=(unlist(random_trees_100k_rf["func"]) <= phangorn::RF.dist(host_tree,tree_func_host_full)) %>% mean,
funcpa_full=(unlist(random_trees_100k_rf["funcpa"]) <= phangorn::RF.dist(host_tree,tree_funcpa_host_full)) %>% mean
)

```

### Panel C

```{r, fig.height=11.5, fig.width=6.5}
### Suppl. Table S2
#phylosym_pvals = data.frame(all=t(p_phylosym), no_ger=t(p_phylosym_noger),no_ptt=t(p_phylosym_noptt), row.names=c("CLR","UniFrac w.", "UniFrac unw.","Jaccard","Func. (CLR)"))
phylosym_pvals = data.frame(all=t(p_phylosym), row.names=c("CLR","UniFrac w.", "UniFrac unw.","Jaccard","Func. (CLR)", "Func. (P/A)"))

write.table(phylosym_pvals %>% rownames_to_column("distance"), "figures_and_tables/Suppl_Table_S4_Phylsymbiosis_pvals.tsv", sep="\t", row.names = F)

panel_c_host = host_tree %>% ggtree(branch.length = "none", size=.8)   + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_c_micro = tree_ufuw_host_full %>% root(node=10, resolve.root=T) %>% ape::rotate(8) %>% as.phylo() %>% ggtree(branch.length = "none", size=.8)  + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_c_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_c_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

###

fig1_panel_c = plot_grid(panel_c_host, panel_c_conn, panel_c_micro, align = "h", ncol=3 )


```

## Fig 1d - sharing analysis

```{r}


sgb_long_red = salmon_rare %>% filter(depth==1000000) %>% group_by(sample, bin) %>% summarise(value=mean(tpm)) %>% left_join(sample_meta) %>% filter(value>0) %>% mutate(host2 = case_when(host == "Ggorilla2" ~ "Ggorilla", host=="Pttrog2" ~ "Pttrog", T ~ host)) %>% ungroup

sgb_share_file = paste0(intermediatedata_folder, "sgb_share.Rds")

if(file.exists(sgb_share_file)){
  sgb_share=readRDS(sgb_share_file)
}else{
sgb_share = lapply(unique(sgb_long_red$host2), function(this_host) {
  lapply(1:100, function(i) {
  sample_subset = sgb_long_red %>% select(host2, sample) %>% distinct() %>% group_by(host2) %>% sample_n(5) %>% pull(sample)
  sgb_this_host = sgb_long_red %>% filter(host2 == this_host, sample %in% sample_subset) %>% pull(bin) %>% unique
  sgb_long_red %>% filter(bin %in% (sgb_this_host), sample %in% sample_subset) %>% 
    select(host2, bin) %>% distinct() %>% group_by(host2) %>% summarise(n=n())  %>%
    left_join(sgb_long_red %>% select(host2) %>% distinct, .) %>% 
    mutate(n_ref = length(sgb_this_host), n = ifelse(is.na(n),0,n), n_rel = n/n_ref) }
  ) %>%  do.call("bind_rows",.) %>% group_by(host2) %>% 
    summarise(mean_n = mean(n_rel)) %>% mutate(ref_host = this_host)
}) %>% do.call("bind_rows",.)
saveRDS(sgb_share, sgb_share_file)
}

sgb_share_rand_file = paste0(intermediatedata_folder, "sgb_share_rand.Rds")

if(file.exists(sgb_share_rand_file)){
  sgb_share_rand=readRDS(sgb_share_rand_file)
}else{
sgb_share_rand = lapply(unique(sgb_long_red$host2), function(this_host) {
  mclapply(1:1000, function(i) {
  sample_rand = sgb_long_red %>% select(host2, sample) %>% distinct() %>% group_by(host2) %>% sample_n(5) %>% pull(sample) %>% sample(5)
  sample_subset = sgb_long_red %>% select(host2, sample) %>% distinct() %>% group_by(host2) %>% sample_n(5) %>% pull(sample)
  sgb_this_host = sgb_long_red %>% filter(sample %in% sample_rand) %>% pull(bin) %>% unique
  sgb_long_red %>% filter(bin %in% (sgb_this_host), sample %in% sample_subset) %>% select(host2, bin) %>% distinct() %>% group_by(host2) %>% summarise(n=n())  %>%
    left_join(sgb_long_red %>% select(host2) %>% distinct, .)  %>% mutate(n_ref = length(sgb_this_host), n_rel = n/n_ref, i = i) 
  }, mc.cores=10) %>% do.call("bind_rows",.)%>% mutate(ref_host = this_host)
}) %>% do.call("bind_rows",.)
saveRDS(sgb_share_rand, sgb_share_rand_file)
}

panel_d_tree = ggtree(host_tree3, size=1)  %>% ggtree::rotate(16) %>% ggtree::rotate(17) 
panel_d_tree$data = panel_d_tree$data %>% left_join(colset %>% mutate(label = case_when(host=="HsA1" ~ "Homo_sapiensA", host=="HsA2" ~ "Homo_sapiensB", host=="HsD" ~ "Homo_sapiensC", T ~ host_tree)))
panel_d_tree = panel_d_tree + geom_tippoint(aes(col=host, shape=host), size=3) + scale_color_manual(values=colset$col, breaks=colset$host) +
     scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host) + theme_nothing() + theme(legend.position = "none") + scale_x_continuous(expand=c(0.5,.5))

panel_d_heatmap=sgb_share_rand %>% filter(!is.na(n_rel)) %>% left_join(sgb_share) %>% mutate(gt = n_rel <= mean_n, lt = n_rel >= mean_n) %>% group_by(ref_host, host2) %>% summarise(p_lt=1 - ((sum(lt)))/(n()+1), p_gt=1 -((sum(gt)))/(n()+1)) %>% left_join(sgb_share) %>% 
  ggplot(aes(x=host2, y=ref_host, alpha=mean_n, fill=ref_host)) + geom_tile() + scale_fill_manual(values=colset$col, breaks=colset$host) + scale_alpha(range=c(0,1)) + 
  geom_text(aes(label = ifelse(p_gt < 0.05, paste0(signif(p_gt,1),sprintf("\u2191")), "")), alpha=1, size=2, family="Arial") + 
  geom_text(aes(label = ifelse(p_lt < 0.05, paste0(signif(p_lt,1),sprintf("\u2193")), "")), alpha=1, size=2, family="Arial") + 
  theme_nothing() + theme(legend.position="none")

panel_d_bottom = colset %>% filter(!host %in% c("Ggorilla2","Pttrog2")) %>% ggplot(aes(x=host, y=1, col=host, shape=host)) + geom_point(size=5) + 
  scale_color_manual(values=colset$col, breaks=colset$host) +
  scale_shape_manual(values=c(15:19)[colset$shape], breaks=colset$host) +
  theme_nothing() + theme(legend.position="none")


#host_tree2 = host_tree %>% bind.tip(tree=., tip.label="Homo_sapiensA", where=which(host_tree$tip.label=="Homo_sapiens")) %>% bind.tip(tree=., tip.label="Homo_sapiensB", where=which(.$tip.label=="Homo_sapiensA"))

fig1_panel_d=plot_grid(plot_grid(panel_d_tree) + theme(plot.margin = unit(c(0,0,0,.5),"cm")), panel_d_heatmap+ theme(plot.margin = unit(c(0,0,0,0),"cm")), NULL, panel_d_bottom + theme(plot.margin = unit(c(0,0,0,0),"cm") ), rel_widths=c(.3,.7), rel_heights = c(.95,.1)) 


```


# Combined for Figure 1 a, b, c, d

```{r, fig.height=11.5, fig.width=6.5}
fig1_left = cowplot::plot_grid(fig1_panel_a + theme(plot.margin = unit(c(1, 0.4, 0, 0), "cm")),
                               fig1_panel_b+theme(legend.position = "none") + theme(plot.margin = unit(c(1, 0.4, 0, 0), "cm")), 
                               cowplot::plot_grid(
                                  get_legend(fig1_panel_b + guides(colour = guide_legend(ncol = 1, title="",byrow = TRUE), shape=guide_legend(ncol = 1, title="",byrow = TRUE)) + 
                                               theme(legend.spacing.y = unit(0.1, 'cm') ,legend.position = "bottom", legend.justification = "top", legend.text = element_text(size=10), legend.key.size = unit(0.2, "cm"))), 
                                  fig1_panel_c, rel_widths=c(.5,.5), labels=c("","c")),
                               fig1_panel_d+theme(legend.position = "none") + theme(plot.margin = unit(c(1, 0.4, 0, 0), "cm")),
                               ncol=1, rel_heights = c(.2,.3,.2,.3), labels=c("a","b","","d"))

plot(fig1_left)

fig1_legend = get_legend(fig1_panel_b + guides(colour = guide_legend(ncol = 1, title="",byrow = TRUE), shape=guide_legend(ncol = 1, title="",byrow = TRUE)) + theme(legend.spacing.y = unit(0.1, 'cm') ,legend.position = "bottom", legend.justification = "center", legend.text = element_text(size=7), legend.key.size = unit(0.2, "cm")))

fig1_new=cowplot::plot_grid(
    cowplot::plot_grid(fig1_panel_a + theme(axis.title = element_text(size=7), axis.text=element_text(size=6)), 
                       fig1_panel_b +theme(legend.position = "none") + theme(axis.title = element_text(size=7), axis.text=element_text(size=6)), 
                       fig1_legend, ncol=3, labels=c("a","b"), rel_widths = c(.3,.4,.3)),
    cowplot::plot_grid(fig1_panel_c, fig1_panel_d, ncol=2, labels=c("c","d"), rel_widths=c(.4,.6)),
    ncol=1
    )

ggsave(filename = "figures_and_tables/Figure_1_new.png",fig1_new, width = 10, height=7.5, bg="white")

library(Cairo)

ggsave(filename = "figures_and_tables/Figure_1_pub.pdf",fig1_new, width = 180, height=130, units = "mm", bg="white", device=cairo_pdf)


grDevices::cairo_pdf("example.pdf")
print(fig1_new)
dev.off()

```

## Source Data

```{r}
library("openxlsx")
# Write the first data set in a new workbook


fig1a_ext = fig1_panel_a$data %>% select(sample, host_long, PhyloDiv = PD)
fig1b_ext = fig1_panel_b$data %>% select(sample, host_long, PCoA1=X1, PCoA2=X2)
fig1d_ext = panel_d_heatmap$data %>% select(-.group)

wb = createWorkbook()

addWorksheet(wb, "Figure1a")
writeData(wb, sheet = "Figure1a", fig1a_ext)

addWorksheet(wb, "Figure1b")
writeData(wb, sheet = "Figure1b", fig1b_ext)

addWorksheet(wb, "Figure1d")
writeData(wb, sheet = "Figure1d", fig1d_ext)

saveRDS(wb, "SourceData_Figure1.Rds")


```



# Combined Figure S2

```{r, fig.height=20, fig.width=25}

### weighted unifrac

panel_uf_micro = tree_ufw_host_full %>% root(node=9, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_uf_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_uf_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_uf_tree = plot_grid(panel_c_host, panel_uf_conn, panel_uf_micro, align = "h", ncol=3 )

                   
                   
### unweighted unifrac

panel_ufuw_micro = tree_ufuw_host_full %>% root(node=10, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_ufuw_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_ufuw_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_ufuw_tree = plot_grid(panel_c_host, panel_ufuw_conn, panel_ufuw_micro, align = "h", ncol=3 )

   
### Gen CLR

panel_clr_micro = tree_clr_host_full %>% root(node=9, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_clr_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_clr_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_clr_tree = plot_grid(panel_c_host, panel_clr_conn, panel_clr_micro, align = "h", ncol=3 )


### Gen Jaccard

panel_jac_micro = tree_jac_host_full %>% root(node=9, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_jac_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_jac_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_jac_tree = plot_grid(panel_c_host, panel_jac_conn, panel_jac_micro, align = "h", ncol=3 )


### function 

panel_func_micro = tree_func_host_full %>% root(node=10, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_func_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_func_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_func_tree = plot_grid(panel_c_host, panel_func_conn, panel_func_micro, align = "h", ncol=3 )

### function pa

panel_funcpa_micro = tree_funcpa_host_full %>% root(node=9, resolve.root=T) %>% ggtree(branch.length = "none", size=.8) + scale_x_reverse() + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5)) + geom_tippoint(data = . %>% left_join(sample_meta, by=c("label"="host_tree")), aes(col=host_long, shape=host_long), size=3) + scale_color_manual(values=colset$col, breaks=colset$host_long) + scale_shape_manual(values=c(15:18)[colset$shape], breaks=colset$host_long)

panel_funcpa_conn = panel_c_host$data %>% filter(isTip) %>% select(label,y_host=y) %>% left_join(panel_funcpa_micro$data %>% filter(isTip) %>% select(label,y_micro=y)) %>% left_join(colset, by=c("label"="host_tree")) %>% ggplot(aes(x=1, xend=2, y=y_host, yend=y_micro)) + geom_segment(size=.8, lty=2) + theme_nothing() + scale_y_continuous(limits=c(0.5, 7.5))

figs2_funcpa_tree = plot_grid(panel_c_host, panel_funcpa_conn, panel_funcpa_micro, align = "h", ncol=3 )

header_plot = plot_grid(ggplot(data=data.frame(label="Ordination")) + geom_text(x=0,y=0,aes(label=label), size=10) + scale_y_continuous(expand=c(0,0), limit=c(-1,1)) + scale_x_continuous(expand=c(0,0), limit=c(-1,1)) + theme_nothing(),
                        ggplot(data=data.frame(label="All Hosts")) + geom_text(x=0,y=0,aes(label=label), size=10) + scale_y_continuous(expand=c(0,0), limit=c(-1,1)) + scale_x_continuous(expand=c(0,0), limit=c(-1,1)) + theme_nothing(),
                        ggplot(data=data.frame(label="P-values")) + geom_text(x=0,y=0,aes(label=label), size=10) + scale_y_continuous(expand=c(0,0), limit=c(-1,1)) + scale_x_continuous(expand=c(0,0), limit=c(-1,1)) + theme_nothing(), ncol=3, rel_widths=c(.4,.3,.2))

fig_s2=plot_grid(header_plot,
cowplot::plot_grid(ufplot1,figs2_uf_tree,phylosym_pvals[2,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
cowplot::plot_grid(ufplot2,figs2_ufuw_tree,phylosym_pvals[3,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
cowplot::plot_grid(clrplot,figs2_clr_tree,phylosym_pvals[1,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
cowplot::plot_grid(jacplot,figs2_jac_tree,phylosym_pvals[4,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
cowplot::plot_grid(funcplot1,figs2_func_tree,phylosym_pvals[5,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
cowplot::plot_grid(funcplot2,figs2_funcpa_tree,phylosym_pvals[6,] %>% gridExtra::tableGrob(), ncol=3, rel_widths=c(.4,.3,.2)),
ncol=1, rel_heights = c(.1,.2,.2,.2,.2,.2,.2))

plot(fig_s2)

ggsave(filename = "figures_and_tables/Figure_S2.png",fig_s2, width = 18, height=20, bg="white")

```

